Generate critical css file for blank theme

mrtuvn opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Generate critical css following below guide for blank theme
https://devdocs.magento.com/guides/v2.4/frontend-dev-guide/css-topics/css-critical-path.html
Default magento only have critical css for luma theme

Critical css is one of factor improvement page speed render. For browser render content above the fold as soon as possible


> How to generate css critical
We have some ways for generate critical 
One choice we can use this repo
But required enviroment setup NodeJS/Npm/Gulp

https://github.com/bobmotor/magento-2-gulp 
This repo current version up to 1.5.0 - used gulp version 4
=> File critical.css output ~26kB

Step following https://github.com/bobmotor/magento-2-gulp/blob/master/README.md
At the last step we use lighthouse for render output critical.css for particular url page (/home, category, details)
We also need some tweak change (clear some declare with font, exclude styles related with email) , lastly modify critical styles following rule
-File size should not too big for block render page (Idealy ~20-50Kb)
-Styles for critical should not overwrite styles declared later (styles-m, styles-l)


### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#<issue_number>

No issue available

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Enable critical path in backend. Or setup via cli command
bin/magento config:set dev/css/use_css_critical_path 1
2. Make sure set blank theme as default in admin 
3. Clear full_page and config cache
4. Go to homepage view page source
5. See critical loaded in head tag of DOM
6. Reload various pages (home, category, details product)  without broken styles visual

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->
CC: @vzabaznov or @adifucan Feel free to give me any advices
@Leland can you help me review this ? Thanks

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#31527: Generate critical css file for blank theme

mrtuvn commented:
@magento run all tests

mrtuvn commented:
rebased mainline code with PR branch

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests CE

sidolov commented:
@magento create issue

ihor-sviziev commented:
Hi @mrtuvn,
Could you explain how you generated this critical.css file so that people can do the same in their own themes?

mrtuvn commented:
Updated in first line PR

hostep commented:
I think we need to understand how luma's critical css was generated, that process might have been different than what that gulp package does.

It would be nice if that approach could be documented somewhere.
Based on the [history](https://github.com/magento/magento2/commits/1f4051381dbdd04658cbc454cdcdff16f191c08b/app/design/frontend/Magento/luma/web/css/critical.css) of that file we could use some information from probably @vzabaznov or @adifucan

adifucan commented:
I do not remember exactly what tool was used to create critical path css for Luma theme. Probably it was either https://www.npmjs.com/package/critical or https://www.npmjs.com/package/penthouse. Or even https://www.sitelocity.com/critical-path-css-generator
But I remember that generated critical path needed to be refactored as well because of some layout problems in generated css.

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests EE

mrtuvn commented:
@magento run Functional Tests EE

mrtuvn commented:
@magento run Functional Tests EE

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8904](https://jira.corp.magento.com/browse/ENGCOM-8904) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

ihor-sviziev commented:
@mrtuvn any info on why you converted this PR to draft?

mrtuvn commented:
I think my solution not completely coverage all real case. So i need to re-check it again (mobile behavior, tablet, desktop..)

ihor-sviziev commented on app/design/frontend/Magento/blank/Magento_Theme/layout/default_head_blocks.xml:
```diff
@@ -7,9 +7,12 @@
 <page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
     <head>
+        <meta name="format-detection" content="telephone=no"/>
+        <font src="fonts/Blank-Theme-Icons/Blank-Theme-Icons.woff2"/>
+        <font src="fonts/opensans/light/opensans-300.woff2"/>
+        <font src="fonts/opensans/regular/opensans-400.woff2"/>
```
I believe it will also appear in the luma theme. Are they needed here? If no - we need to remove them

mrtuvn replied:
i'm not sure why it's moved to luma. But in blank theme we will not get benefit of preload like luma font defined

mrtuvn replied:
Everything ready for review! My main purpose add font here for blank to allow browser preload font that i think it's need for render properly. It's prevent layout shifting and flashes of invisible text. Only keep lowest request http as possible. I don't add all font here like luma did

ihor-sviziev replied:
Adding some font to a blank theme - you're adding it to all children themes by default, for instance, to luma. We need to double-check that.

mrtuvn replied:
ok let's revert this back to previous!

ihor-sviziev replied:
Actually we can keep it and add remove these fonts from the luma theme (as it inherits them), but that's not actually related to changes in critical.css

ihor-sviziev reviewed (commented):
Have you finished? Can I review?

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run Functional Tests B2B, WebAPI Tests

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8904](https://jira.corp.magento.com/browse/ENGCOM-8904) has been created to process this Pull Request

mrtuvn commented:
Update with styles loader by css instead use image

mrtuvn commented:
@magento give me test instance

mrtuvn committed:
Generate critical css file for blank theme


update


update


update


update


update


Update


update


revert change in layout


update


update


update


update


update

ihor-sviziev commented:
Hi @mrtuvn,
I just reviewed this PR from another perspective.
It feels like there no sense in providing the critical.css file for the blank theme, as it was designed to be just a "base" theme, and I don't believe someone will run it in production.
It makes sense to provide such a file for themes used on the frontend, in case of default Magento installation - that's the Luma theme.

If it makes sense to you - I suggest closing this PR. What do you think? 🤔

mrtuvn commented:
Yes i think this solution depends on each store build by merchant. Each store have difference way and theme so it's seem no base answer for all of them. They will have resources for build critical for they own site instead want to use from blank

ihor-sviziev commented:
@mrtuvn i would prefer to close it
